README.md
# Plugin WooCommerce pour constituer des packs de bouteilles

Ce plugin permet d'appliquer des réductions automatiquement au panier dès que des packs de 3, 6 ou 12 unités sont constitués. A été développé sur mesure pour la [brasserie Walpine](https://www.walpine.fr) et leurs spécificités liées au logiciel de gestion de brasserie artisanale [SuperG](https://www.superg.fr).

Les packs peuvent être constitués de plusieurs bières différentes, chacune ayant son SKU/UGS dédié pour facilité la gestion des commandes par la suite lors de la synchronisation avec [SuperG](https://www.superg.fr).

![Exemple de réduction appliqué à un panier de 6 bouteilles, 4 blondes, 1 IPA et 1 blanche](https://depot.studiomaiis.net/screenshots/walpine_panier.png "Exemple de réduction appliqué à un panier de 6 bouteilles, 4 blondes, 1 IPA et 1 blanche")

## Configuration requise

* PHP : 7.4 et plus
* WordPress : 5.6.9 et plus
* WooCommerce : 5.8 et plus
* Plugin ACF Pro : 5 et plus, requis car la grille tarifaire dépend du type de champ "Repéteur"

Il se peut que le plugin fonctionne sur des versions antérieures de PHP, WordPress ou WooCommerce, mais il n'a pas été testé dans d'autres conditions que celles mentionnées ci-dessus.

## Installation & mises à jour

La première installation est manuelle. [Téléchargez l'archive ici](https://depot.studiomaiis.net/wordpress/woocommerce-coupon-packs.zip) et installez le plugin de manière standard :
* via FTP une fois l'archive décompressée dans `wp-content/plugins/`,
* ou via la page d'administration des extensions.

Si vous téléchargez le code depuis Github, renommer le répertoire `wp-plugin-woocommerce-coupon-packs(*)` en `woocommerce-coupon-packs`.

Une fois le module installé, les mises à jour se font depuis le back-office de WordPress.

## Configuration

1. Allez dans *Marketing > Codes promo* et créez un nouveau code promo. Peut importe son nom mais *REMISE_PACKS* semble approprié pour une meilleure identification dans les commandes,

2. Sélectionnez *Packs* dans l'onglet *Général > Type de remise*,

3. Activez *Code promo pour la configuration des packs ?*,

4. Saisissez vos tarifs dans la grille.

![Page de configuration d'un code promo de type Packs](https://depot.studiomaiis.net/screenshots/woo_packs_config.png "Page de configuration d'un code promo de type Packs")

**Important** : il ne peut y avoir qu'un seul code promo de type *Packs* ; ne pas utiliser un SKU/UGS sur plusieurs lignes.

## Personnalisations

Ce plugin est assez spécifique à la [brasserie du Champsaur Walpine.fr](https://www.walpine.fr) mais grâce à ACF Pro, et moyennant quelques développements WordPress sur mesure, il peut être rapidement adapté à d'autres besoins.

N'hésitez pas à [me contacter](https://www.studiomaiis.net) pour vous créer un plugin sur mesure !.



CHANGELOG.md
#### Version 0.2

* Tests de mise à jour auto

#### Version 0.1

* Rédaction documentation




includes/acf-packs.php
<?php


if( function_exists('acf_add_local_field_group') ):

acf_add_local_field_group(array(
	'key' => 'group_6246e00d4baf5',
	'title' => 'Packs',
	'fields' => array(
		array(
			'key' => 'field_6246e04df76a7',
			'label' => 'Code promo pour la configuration des packs ?',
			'name' => 'packs',
			'type' => 'true_false',
			'instructions' => '',
			'required' => 0,
			'conditional_logic' => 0,
			'wrapper' => array(
				'width' => '',
				'class' => '',
				'id' => '',
			),
			'message' => '',
			'default_value' => 0,
			'ui' => 1,
			'ui_on_text' => '',
			'ui_off_text' => '',
		),
		array(
			'key' => 'field_6246fe1c5296b',
			'label' => 'Tarifs',
			'name' => 'tarifs',
			'type' => 'repeater',
			'instructions' => '',
			'required' => 1,
			'conditional_logic' => array(
				array(
					array(
						'field' => 'field_6246e04df76a7',
						'operator' => '==',
						'value' => '1',
					),
				),
			),
			'wrapper' => array(
				'width' => '',
				'class' => '',
				'id' => '',
			),
			'collapsed' => '',
			'min' => 1,
			'max' => 0,
			'layout' => 'table',
			'button_label' => 'Ajouter une grille de tarifs',
			'sub_fields' => array(
				array(
					'key' => 'field_624700da52972',
					'label' => 'Titre',
					'name' => 'titre',
					'type' => 'text',
					'instructions' => 'Sera précédé dans le panier par : nombre de pack + "packs de " + nombre d\'unités
Exemple en saissant "classiques 33cl" :
3 packs de 6 classiques 33cl',
					'required' => 1,
					'conditional_logic' => 0,
					'wrapper' => array(
						'width' => '20',
						'class' => '',
						'id' => '',
					),
					'default_value' => '',
					'placeholder' => 'classiques 33cl',
					'prepend' => '',
					'append' => '',
					'maxlength' => '',
				),
				array(
					'key' => 'field_6247001b5296e',
					'label' => 'SKUs / UGS',
					'name' => 'skus',
					'type' => 'text',
					'instructions' => 'La liste des SKUS concernés, séparés par des virgules, sans espace (formatage important)',
					'required' => 1,
					'conditional_logic' => 0,
					'wrapper' => array(
						'width' => '50',
						'class' => '',
						'id' => '',
					),
					'default_value' => '',
					'placeholder' => 'SUPERG_1_1,SUPERG_2_1',
					'prepend' => '',
					'append' => '',
					'maxlength' => '',
				),
				array(
					'key' => 'field_6247007c5296f',
					'label' => 'x3',
					'name' => 'x3',
					'type' => 'number',
					'instructions' => '',
					'required' => 1,
					'conditional_logic' => 0,
					'wrapper' => array(
						'width' => '',
						'class' => '',
						'id' => '',
					),
					'default_value' => 0,
					'placeholder' => '',
					'prepend' => '',
					'append' => '',
					'min' => '',
					'max' => '',
					'step' => '',
				),
				array(
					'key' => 'field_624700ac52970',
					'label' => 'x6',
					'name' => 'x6',
					'type' => 'number',
					'instructions' => '',
					'required' => 1,
					'conditional_logic' => 0,
					'wrapper' => array(
						'width' => '',
						'class' => '',
						'id' => '',
					),
					'default_value' => 0,
					'placeholder' => '',
					'prepend' => '',
					'append' => '',
					'min' => '',
					'max' => '',
					'step' => '',
				),
				array(
					'key' => 'field_624700c252971',
					'label' => 'x12',
					'name' => 'x12',
					'type' => 'number',
					'instructions' => '',
					'required' => 1,
					'conditional_logic' => 0,
					'wrapper' => array(
						'width' => '',
						'class' => '',
						'id' => '',
					),
					'default_value' => 0,
					'placeholder' => '',
					'prepend' => '',
					'append' => '',
					'min' => '',
					'max' => '',
					'step' => '',
				),
			),
		),
	),
	'location' => array(
		array(
			array(
				'param' => 'post_type',
				'operator' => '==',
				'value' => 'shop_coupon',
			),
		),
	),
	'menu_order' => 0,
	'position' => 'normal',
	'style' => 'seamless',
	'label_placement' => 'top',
	'instruction_placement' => 'label',
	'hide_on_screen' => '',
	'active' => true,
	'description' => '',
	'show_in_rest' => 0,
));

endif;



woocommerce-coupon-packs.php
<?php

/**
 * Plugin Name: WooCommerce Réduction par packs
 * Description: Applique automatiquement des réductions dans le panier dès que des packs sont constitués
 * Plugin URI: https://github.com/studiomaiis/wp-plugin-woocommerce-coupon-packs
 * Version: 0.2
 * Requires at least: 5.6
 * Tested up to: 6.0
 * Requires PHP: 7.3
 * Author: Pierre BASSON
 * Author URI: https://www.studiomaiis.net
 * Text Domain: woocommerce-coupon-packs
 * Domain Path: /languages
 *
 * WC requires at least: 5.8
 * WC tested up to: 6.9.4
 *
 * License: GNU General Public License v3.0
 * License URI: https://www.gnu.org/licenses/gpl-3.0.html
 */

if ( ! defined( 'ABSPATH' ) ) {
	exit;
}

/**
 * Required minimums and constants
 */
define( 'WC_COUPON_PACKS_SERVER_URL', 'https://depot.studiomaiis.net/wordpress' );
define( 'WC_COUPON_PACKS_PREFIX', 'wc_coupon_packs' );
define( 'WC_COUPON_PACKS_PLUGIN_NAME', __( 'WooCommerce Réduction par packs', 'woocommerce-coupon-packs') );
define( 'WC_COUPON_PACKS_MAIN_FILE', __FILE__ );
define( 'WC_COUPON_PACKS_ABSPATH', __DIR__ . '/' );
define( 'WC_COUPON_PACKS_PLUGIN_URL', untrailingslashit( plugins_url( basename( plugin_dir_path( __FILE__ ) ), basename( __FILE__ ) ) ) );
define( 'WC_COUPON_PACKS_PLUGIN_PATH', untrailingslashit( plugin_dir_path( __FILE__ ) ) );


function woocommerce_coupon_packs() {


	static $plugin;


	if ( ! isset( $plugin ) ) {


		class WC_Coupon_Packs {


			/**
			 * The *Singleton* instance of this class
			 *
			 * @var Singleton
			 */
			private static $instance;


			/**
			 * Returns the *Singleton* instance of this class.
			 *
			 * @return Singleton The *Singleton* instance.
			 */
			public static function get_instance() {
				if ( null === self::$instance ) {
					self::$instance = new self();
				}
				return self::$instance;
			}


			/**
			 * The main Mercanet gateway instance. Use get_main_mercanet_gateway() to access it.
			 *
			 * @var null|WC_Mercanet_Payment_Gateway
			 */
			protected $gateway = null;


			/**
			 * Private clone method to prevent cloning of the instance of the
			 * *Singleton* instance.
			 *
			 * @return void
			 */
			public function __clone() {}


			/**
			 * Private unserialize method to prevent unserializing of the *Singleton*
			 * instance.
			 *
			 * @return void
			 */
			public function __wakeup() {}


			/**
			 * Protected constructor to prevent creating a new instance of the
			 * *Singleton* via the `new` operator from outside of this class.
			 */
			public function __construct() {
				add_action( 'admin_init', [ $this, 'install' ] );
				
				$this->init();
			}


			/**
			 * Init the plugin after plugins_loaded so environment variables are set.
			 */
			public function init() {
				require_once dirname( __FILE__ ) . '/includes/class-packs-updater.php';
				require_once dirname( __FILE__ ) . '/includes/acf-packs.php';
				
				$updater = new Packs\Updater( __FILE__, WC_COUPON_PACKS_SERVER_URL );
				$updater->initialize();

				add_filter( 'woocommerce_coupon_discount_types', [ $this, 'coupon_discount_types' ] );
				add_action( 'woocommerce_before_cart', [ $this, 'before_checkout_form' ] );
				add_action( 'woocommerce_before_checkout_form', [ $this, 'before_checkout_form' ] );
				add_filter( 'woocommerce_coupon_custom_discounts_array', [ $this, 'coupon_custom_discounts_array' ], 10, 2 );
				add_filter( 'woocommerce_coupon_message', [ $this, 'coupon_message' ], 10, 3 );
				add_filter( 'woocommerce_coupon_error', [ $this, 'coupon_message' ], 10, 3 );
				add_filter( 'woocommerce_cart_totals_coupon_html', [ $this, 'cart_totals_coupon_html' ], 10, 3 );
				add_filter( 'woocommerce_cart_totals_coupon_label', [ $this, 'cart_totals_coupon_label' ], 10, 2 );
			}


			public function update_plugin_version() {
				delete_option( WC_COUPON_PACKS_PREFIX . '_version' );

				if( ! function_exists('get_plugin_data') ){
					require_once( ABSPATH . 'wp-admin/includes/plugin.php' );
				}
				$plugin_data = get_plugin_data( __FILE__ );

				update_option( WC_COUPON_PACKS_PREFIX . '_version', $plugin_data['Version'] );
			}


			public function install() {
				if ( ! is_plugin_active( plugin_basename( __FILE__ ) ) ) {
					return;
				}
	
				$this->update_plugin_version();
			}


			public function coupon_discount_types( $discount_types ) {
				$discount_types['packs'] = "Packs";
				
				return $discount_types;
			}
			
			
			public function before_checkout_form() {
			
				if ( is_admin() and ! defined( 'DOING_AJAX' ) ) {
					return;
				}
				
				$posts = get_posts( array(
					'post_type' => 'shop_coupon',
					'posts_per_page' => 1,
					'meta_query' => array(
						array(
							'key' => 'discount_type',
							'value' => 'packs',
						),
						array(
							'key' => 'packs',
							'value' => '1',
						),
					),
				));
				
				if ( count( $posts ) == 1 ) {
					
					$coupon = reset( $posts );

					if ( WC()->cart->get_subtotal() > 0 ) {
						if ( ! in_array( $coupon->post_title, WC()->cart->get_applied_coupons() ) ) {
							WC()->cart->apply_coupon( $coupon->post_title );
						}
					} else {
						WC()->cart->remove_coupon( $coupon->post_title );
					}
				
				}
			}
			
			
			public function coupon_custom_discounts_array( $discounts, $coupon ) {
				
				$cart = WC()->cart;
			
				$reductions_packs = self::process_packs_discounts( $cart );
				
				if ( $reductions_packs[ 'montant' ] > 0 ) {
				
					$discounts[ array_key_first( $discounts ) ] = $reductions_packs[ 'montant' ] * 100;
					
				}
			
				return $discounts;
			}
			
			
			public function coupon_message( $msg, $msg_code, $coupon ) {
				if ( $coupon->get_discount_type() == 'packs' ) {
					$msg = '';
				}
				
				return $msg;
			}


			public function cart_totals_coupon_html( $coupon_html, $coupon, $discount_amount_html ) {
				
				if ( $coupon->get_discount_type() == 'packs' ) {
					$cart = WC()->cart;
					$reductions_packs = self::process_packs_discounts( $cart );
					if ( $reductions_packs[ 'montant' ] == 0 ) {
						$coupon_html = 'Aucun pour le moment';
					} else {
						$coupon_html = $discount_amount_html;
						if ( count( $reductions_packs[ 'detail' ] ) ) {
							$coupon_html .= '<br>' . join( '<br>', $reductions_packs[ 'detail' ] );
						}
					}
				}
				
				return $coupon_html;
				
			}
			
			
			public function cart_totals_coupon_label( $label, $coupon ) {
				if ( $coupon->get_discount_type() == 'packs' ) {
					$label = "Packs constitués";
				}
				
				return $label;
			}

			
			public function process_packs_discounts( $cart ) {
			
				$reductions_packs = array(
					'montant' => 0,
					'detail' => array(),
				);
				$quantites = $groupes = array();
				
				$posts = get_posts(array(
					'post_type' => 'shop_coupon',
					'posts_per_page' => 1,
					'meta_query' => array(
						array(
							'key' => 'discount_type',
							'value' => 'packs',
						),
						array(
							'key' => 'packs',
							'value' => '1',
						),
					),
				));
				
				if (count($posts) == 1) {
			
					if (count($cart->cart_contents)) {
						foreach($cart->cart_contents as $item) {
							$sku = $item['data']->get_sku();
							$quantite = $item['quantity'];
							if (!isset($quantites[$sku])) $quantites[$sku] = array('quantite' => $quantite, 'prix_unitaire' => intval($item['data']->get_price() * 100));
							else $quantites[$sku]['quantite']+= $quantite;
						}
					}
			
					$post_id = $posts[0]->ID;
					$tarifs = get_field('tarifs', $post_id);
					$i = 1;
					
					foreach ($tarifs as $tarif) {
						$groupes[$i] = array(
							'quantite' => 0,
						);
						$skus = explode(',', $tarif['skus']);
						foreach ($skus as $sku) {
							if (isset($quantites[$sku])) {
								$groupes[$i]['quantite']+= $quantites[$sku]['quantite'];
								$prix_unitaire = $quantites[$sku]['prix_unitaire'];
								if (isset($groupes[$i]['prix_unitaire'])) {
									$groupes[$i]['prix_unitaire'] = max($prix_unitaire, $groupes[$i]['prix_unitaire']);
								} else {
									$groupes[$i]['prix_unitaire'] = $prix_unitaire;
								}
							}
						}
						if ($groupes[$i]['quantite'] > 0) {
							$reduction = self::process_packs_discount($tarif, $groupes[$i]['prix_unitaire'], $groupes[$i]['quantite']);
							if ($reduction['montant'] > 0) {
								$reductions_packs['montant']+= $reduction['montant'];
								$reductions_packs['detail'][] = join(', ', $reduction['detail']);
							}
						}
						$i++;
					}
			
				}
			
				return $reductions_packs;
			}
			
			
			public function process_packs_discount( $tarif, $prix_unitaire, $quantite ) {
			
				$reduction = array(
					'montant' => 0,
					'detail' => array(),
				);
				$packs = array();
				
				foreach (array(12, 6, 3) as $nb_unites) {
					$x = intdiv($quantite, $nb_unites);
					if ($x > 0) {
						$reduction_1_pack = ($prix_unitaire / 100) * $nb_unites - $tarif['x'.$nb_unites];
						$reduction['montant']+= $x * $reduction_1_pack;
						$reduction['detail'][] = $x.' pack'.($x > 1 ? 's' : '').' de '.$nb_unites.' '.$tarif['titre'];
					}
					$reste = fmod($quantite, $nb_unites);
					if ($reste > 0) {
						$quantite = $reste;
					} else {
						break;
					}
				}
			
				return $reduction;
			}

		}


		$plugin = WC_Coupon_Packs::get_instance();


	}
	
	return $plugin;

}


add_action( 'plugins_loaded', 'woocommerce_coupon_packs_init' );
function woocommerce_coupon_packs_init() {
	load_plugin_textdomain( 'woocommerce-coupon-packs', false, dirname( plugin_basename( __FILE__ ) ) . '/languages' );

	if ( ! class_exists( 'WooCommerce' ) ) {
		add_action( 'admin_notices', 'woocommerce_coupon_packs_missing_wc_notice' );
		return;
	}

	if ( ! class_exists( 'acf_pro' ) ) {
		add_action( 'admin_notices', 'woocommerce_coupon_packs_missing_acf_pro_notice' );
		return;
	}

	woocommerce_coupon_packs();
}


function woocommerce_coupon_packs_missing_wc_notice() {
	/* translators: 1. Plugin name 2. URL link. */
	echo '<div class="error"><p><strong>' . sprintf( esc_html__( '%1$s requires WooCommerce to be installed and active. You can download %2$s here.', 'woocommerce-coupon-packs' ), WC_COUPON_PACKS_PLUGIN_NAME, '<a href="https://woocommerce.com/" target="_blank">WooCommerce</a>' ) . '</strong></p></div>';
}


function woocommerce_coupon_packs_missing_acf_pro_notice() {
	/* translators: 1. Plugin name */
	echo '<div class="error"><p><strong>' . sprintf( esc_html__( '%1$s requires ACF Pro to be installed and active.', 'woocommerce-coupon-packs' ), WC_COUPON_PACKS_PLUGIN_NAME ) . '</strong></p></div>';
}



includes/class-packs-updater.php
<?php


namespace Packs;


if ( ! defined( 'ABSPATH' ) ) {
	exit;
}


class Updater {

	private $file;
	private $server_url;
	private $plugin;
	private $basename;
	private $active;
	private $server_response;

	public function __construct( $file, $server_url ) {

		$this->file = $file;
		$this->server_url = $server_url;

		add_action( 'admin_init', array( $this, 'set_plugin_properties' ) );

		return $this;
	}

	public function set_plugin_properties() {
		$this->plugin	= get_plugin_data( $this->file );
		$this->basename = plugin_basename( $this->file );
		$this->active	= is_plugin_active( $this->basename );
	}

	private function get_server_info() {
		$this->set_plugin_properties();
		
		if ( is_null( $this->server_response ) ) {
			$slug = current( explode( '/' , $this->basename ) );
			$request_uri = $this->server_url.'/'.$slug.'.json';
			
			$response = json_decode( wp_remote_retrieve_body( wp_remote_get( $request_uri ) ), true ); // Get JSON and parse it

			$this->server_response = $response;
		}
	}

	public function initialize() {
		add_filter( 'pre_set_site_transient_update_plugins', array( $this, 'modify_transient' ), 10, 1 );
		add_filter( 'plugins_api', array( $this, 'plugin_popup' ), 10, 3);
		add_filter( 'upgrader_post_install', array( $this, 'after_install' ), 10, 3 );
	}

	public function modify_transient( $transient ) {

		if ( property_exists( $transient, 'checked') ) {

			if ( $checked = $transient->checked ) {

				$this->get_server_info();
				
				
				$out_of_date = version_compare( $this->server_response['version'], $checked[ $this->basename ], 'gt' ); // Check if we're out of date

				if ( $out_of_date ) {
					
					$slug = current( explode( '/' , $this->basename ) );
					
					$plugin = array(
						'url' => $this->plugin["PluginURI"],
						'slug' => $slug,
						'package' => $this->server_response['download_url'],
						'new_version' => $this->server_response['version']
					);
					
					$transient->response[$this->basename] = (object) $plugin;

				}
			}
		}

		return $transient;
	}

	public function plugin_popup( $result, $action, $args ) {

		if ( ! empty( $args->slug ) ) {
			
			if ( $args->slug == current( explode( '/' , $this->basename ) ) ) {
				
				$this->get_server_info();
				
				return (object) $this->server_response;

			}

		}
		
		return $result;
	}
	
	public function after_install( $response, $hook_extra, $result ) {
		global $wp_filesystem; // Get global FS object

		$install_directory = plugin_dir_path( $this->file ); // Our plugin directory
		$wp_filesystem->move( $result['destination'], $install_directory ); // Move files to the plugin dir
		$result['destination'] = $install_directory; // Set the destination for the rest of the stack

		if ( $this->active ) { // If it was active
			activate_plugin( $this->basename ); // Reactivate
		}

		return $result;
	}

}



woocommerce-packs-bouteilles.php
<?php

/**
 * Plugin Name: WooCommerce Réduction par packs
 * Description: Applique automatiquement des réductions dans le panier dès que des packs sont constitués
 * Plugin URI: https://github.com/studiomaiis/wp-plugin-woocommerce-coupon-packs
 * Version: 0.1
 * Requires at least: 5.6
 * Tested up to: 6.0
 * Requires PHP: 7.3
 * Author: Pierre BASSON
 * Author URI: https://www.studiomaiis.net
 * Text Domain: woocommerce-coupon-packs
 * Domain Path: /languages
 *
 * WC requires at least: 5.8
 * WC tested up to: 6.9.4
 *
 * License: GNU General Public License v3.0
 * License URI: https://www.gnu.org/licenses/gpl-3.0.html
 */

if ( ! defined( 'ABSPATH' ) ) {
	exit;
}

/**
 * Required minimums and constants
 */
define( 'WC_COUPON_PACKS_SERVER_URL', 'https://depot.studiomaiis.net/wordpress' );
define( 'WC_COUPON_PACKS_PREFIX', 'wc_coupon_packs' );
define( 'WC_COUPON_PACKS_PLUGIN_NAME', __( 'WooCommerce Réduction par packs', 'woocommerce-coupon-packs') );
define( 'WC_COUPON_PACKS_MAIN_FILE', __FILE__ );
define( 'WC_COUPON_PACKS_ABSPATH', __DIR__ . '/' );
define( 'WC_COUPON_PACKS_PLUGIN_URL', untrailingslashit( plugins_url( basename( plugin_dir_path( __FILE__ ) ), basename( __FILE__ ) ) ) );
define( 'WC_COUPON_PACKS_PLUGIN_PATH', untrailingslashit( plugin_dir_path( __FILE__ ) ) );


function woocommerce_coupon_packs() {


	static $plugin;


	if ( ! isset( $plugin ) ) {


		class WC_Coupon_Packs {


			/**
			 * The *Singleton* instance of this class
			 *
			 * @var Singleton
			 */
			private static $instance;


			/**
			 * Returns the *Singleton* instance of this class.
			 *
			 * @return Singleton The *Singleton* instance.
			 */
			public static function get_instance() {
				if ( null === self::$instance ) {
					self::$instance = new self();
				}
				return self::$instance;
			}


			/**
			 * The main Mercanet gateway instance. Use get_main_mercanet_gateway() to access it.
			 *
			 * @var null|WC_Mercanet_Payment_Gateway
			 */
			protected $gateway = null;


			/**
			 * Private clone method to prevent cloning of the instance of the
			 * *Singleton* instance.
			 *
			 * @return void
			 */
			public function __clone() {}


			/**
			 * Private unserialize method to prevent unserializing of the *Singleton*
			 * instance.
			 *
			 * @return void
			 */
			public function __wakeup() {}


			/**
			 * Protected constructor to prevent creating a new instance of the
			 * *Singleton* via the `new` operator from outside of this class.
			 */
			public function __construct() {
				add_action( 'admin_init', [ $this, 'install' ] );
				
				$this->init();
			}


			/**
			 * Init the plugin after plugins_loaded so environment variables are set.
			 */
			public function init() {
				require_once dirname( __FILE__ ) . '/includes/class-mercanet-updater.php';
				require_once dirname( __FILE__ ) . '/includes/acf-packs.php';
				
				$updater = new Packs\Updater( __FILE__, WC_COUPON_PACKS_SERVER_URL );
				$updater->initialize();

				add_filter( 'woocommerce_coupon_discount_types', [ $this, 'coupon_discount_types' ] );
				add_action( 'woocommerce_before_cart', [ $this, 'before_checkout_form' ] );
				add_action( 'woocommerce_before_checkout_form', [ $this, 'before_checkout_form' ] );
				add_filter( 'woocommerce_coupon_custom_discounts_array', [ $this, 'coupon_custom_discounts_array' ], 10, 2 );
				add_filter( 'woocommerce_coupon_message', [ $this, 'coupon_message' ], 10, 3 );
				add_filter( 'woocommerce_coupon_error', [ $this, 'coupon_message' ], 10, 3 );
				add_filter( 'woocommerce_cart_totals_coupon_html', [ $this, 'cart_totals_coupon_html' ], 10, 3 );
				add_filter( 'woocommerce_cart_totals_coupon_label', [ $this, 'cart_totals_coupon_label' ], 10, 2 );
			}


			public function update_plugin_version() {
				delete_option( WC_COUPON_PACKS_PREFIX . '_version' );

				if( ! function_exists('get_plugin_data') ){
					require_once( ABSPATH . 'wp-admin/includes/plugin.php' );
				}
				$plugin_data = get_plugin_data( __FILE__ );

				update_option( WC_COUPON_PACKS_PREFIX . '_version', $plugin_data['Version'] );
			}


			public function install() {
				if ( ! is_plugin_active( plugin_basename( __FILE__ ) ) ) {
					return;
				}
	
				$this->update_plugin_version();
			}


			public function coupon_discount_types( $discount_types ) {
				$discount_types['packs'] = "Packs";
				
				return $discount_types;
			}
			
			
			public function before_checkout_form() {
			
				if ( is_admin() and ! defined( 'DOING_AJAX' ) ) {
					return;
				}
				
				$posts = get_posts( array(
					'post_type' => 'shop_coupon',
					'posts_per_page' => 1,
					'meta_query' => array(
						array(
							'key' => 'discount_type',
							'value' => 'packs',
						),
						array(
							'key' => 'packs',
							'value' => '1',
						),
					),
				));
				
				if ( count( $posts ) == 1 ) {
					
					$coupon = reset( $posts );

					if ( WC()->cart->get_subtotal() > 0 ) {
						if ( ! in_array( $coupon->post_title, WC()->cart->get_applied_coupons() ) ) {
							WC()->cart->apply_coupon( $coupon->post_title );
						}
					} else {
						WC()->cart->remove_coupon( $coupon->post_title );
					}
				
				}
			}
			
			
			public function coupon_custom_discounts_array( $discounts, $coupon ) {
				
				$cart = WC()->cart;
			
				$reductions_packs = self::process_packs_discounts( $cart );
				
				if ( $reductions_packs[ 'montant' ] > 0 ) {
				
					$discounts[ array_key_first( $discounts ) ] = $reductions_packs[ 'montant' ] * 100;
					
				}
			
				return $discounts;
			}
			
			
			public function coupon_message( $msg, $msg_code, $coupon ) {
				if ( $coupon->get_discount_type() == 'packs' ) {
					$msg = '';
				}
				
				return $msg;
			}


			public function cart_totals_coupon_html( $coupon_html, $coupon, $discount_amount_html ) {
				
				if ( $coupon->get_discount_type() == 'packs' ) {
					$cart = WC()->cart;
					$reductions_packs = self::process_packs_discounts( $cart );
					if ( $reductions_packs[ 'montant' ] == 0 ) {
						$coupon_html = 'Aucun pour le moment';
					} else {
						$coupon_html = $discount_amount_html;
						if ( count( $reductions_packs[ 'detail' ] ) ) {
							$coupon_html .= '<br>' . join( '<br>', $reductions_packs[ 'detail' ] );
						}
					}
				}
				
				return $coupon_html;
				
			}
			
			
			public function cart_totals_coupon_label( $label, $coupon ) {
				if ( $coupon->get_discount_type() == 'packs' ) {
					$label = "Packs constitués";
				}
				
				return $label;
			}

			
			public function process_packs_discounts( $cart ) {
			
				$reductions_packs = array(
					'montant' => 0,
					'detail' => array(),
				);
				$quantites = $groupes = array();
				
				$posts = get_posts(array(
					'post_type' => 'shop_coupon',
					'posts_per_page' => 1,
					'meta_query' => array(
						array(
							'key' => 'discount_type',
							'value' => 'packs',
						),
						array(
							'key' => 'packs',
							'value' => '1',
						),
					),
				));
				
				if (count($posts) == 1) {
			
					if (count($cart->cart_contents)) {
						foreach($cart->cart_contents as $item) {
							$sku = $item['data']->get_sku();
							$quantite = $item['quantity'];
							if (!isset($quantites[$sku])) $quantites[$sku] = array('quantite' => $quantite, 'prix_unitaire' => intval($item['data']->get_price() * 100));
							else $quantites[$sku]['quantite']+= $quantite;
						}
					}
			
					$post_id = $posts[0]->ID;
					$tarifs = get_field('tarifs', $post_id);
					$i = 1;
					
					foreach ($tarifs as $tarif) {
						$groupes[$i] = array(
							'quantite' => 0,
						);
						$skus = explode(',', $tarif['skus']);
						foreach ($skus as $sku) {
							if (isset($quantites[$sku])) {
								$groupes[$i]['quantite']+= $quantites[$sku]['quantite'];
								$prix_unitaire = $quantites[$sku]['prix_unitaire'];
								if (isset($groupes[$i]['prix_unitaire'])) {
									$groupes[$i]['prix_unitaire'] = max($prix_unitaire, $groupes[$i]['prix_unitaire']);
								} else {
									$groupes[$i]['prix_unitaire'] = $prix_unitaire;
								}
							}
						}
						if ($groupes[$i]['quantite'] > 0) {
							$reduction = self::process_packs_discount($tarif, $groupes[$i]['prix_unitaire'], $groupes[$i]['quantite']);
							if ($reduction['montant'] > 0) {
								$reductions_packs['montant']+= $reduction['montant'];
								$reductions_packs['detail'][] = join(', ', $reduction['detail']);
							}
						}
						$i++;
					}
			
				}
			
				return $reductions_packs;
			}
			
			
			public function process_packs_discount( $tarif, $prix_unitaire, $quantite ) {
			
				$reduction = array(
					'montant' => 0,
					'detail' => array(),
				);
				$packs = array();
				
				foreach (array(12, 6, 3) as $nb_unites) {
					$x = intdiv($quantite, $nb_unites);
					if ($x > 0) {
						$reduction_1_pack = ($prix_unitaire / 100) * $nb_unites - $tarif['x'.$nb_unites];
						$reduction['montant']+= $x * $reduction_1_pack;
						$reduction['detail'][] = $x.' pack'.($x > 1 ? 's' : '').' de '.$nb_unites.' '.$tarif['titre'];
					}
					$reste = fmod($quantite, $nb_unites);
					if ($reste > 0) {
						$quantite = $reste;
					} else {
						break;
					}
				}
			
				return $reduction;
			}

		}


		$plugin = WC_Coupon_Packs::get_instance();


	}
	
	return $plugin;

}


add_action( 'plugins_loaded', 'woocommerce_coupon_packs_init' );
function woocommerce_coupon_packs_init() {
	load_plugin_textdomain( 'woocommerce-coupon-packs', false, dirname( plugin_basename( __FILE__ ) ) . '/languages' );

	if ( ! class_exists( 'WooCommerce' ) ) {
		add_action( 'admin_notices', 'woocommerce_coupon_packs_missing_wc_notice' );
		return;
	}

	if ( ! class_exists( 'acf_pro' ) ) {
		add_action( 'admin_notices', 'woocommerce_coupon_packs_missing_acf_pro_notice' );
		return;
	}

	woocommerce_coupon_packs();
}


function woocommerce_coupon_packs_missing_wc_notice() {
	/* translators: 1. Plugin name 2. URL link. */
	echo '<div class="error"><p><strong>' . sprintf( esc_html__( '%1$s requires WooCommerce to be installed and active. You can download %2$s here.', 'woocommerce-coupon-packs' ), WC_COUPON_PACKS_PLUGIN_NAME, '<a href="https://woocommerce.com/" target="_blank">WooCommerce</a>' ) . '</strong></p></div>';
}


function woocommerce_coupon_packs_missing_acf_pro_notice() {
	/* translators: 1. Plugin name */
	echo '<div class="error"><p><strong>' . sprintf( esc_html__( '%1$s requires ACF Pro to be installed and active.', 'woocommerce-coupon-packs' ), WC_COUPON_PACKS_PLUGIN_NAME ) . '</strong></p></div>';
}

